src/test/scala/nl/knaw/dans/lib/taskqueue/ActiveTaskQueueSpec.scala

```diff
@@ -0,0 +1,126 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package nl.knaw.dans.lib.taskqueue
+
+import org.scalamock.scalatest.MockFactory
+import org.scalatest.OneInstancePerTest
+import org.scalatest.concurrent.{ Eventually, IntegrationPatience }
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers
+
+import scala.util.Success
+
+class ActiveTaskQueueSpec extends AnyFlatSpec with Matchers with OneInstancePerTest with Eventually with IntegrationPatience with MockFactory {
+
+  "start" should "process previously queued tasks" in {
+    val taskQueue = new ActiveTaskQueue[Any]()
+    val triggeredTasks = List(
+      TriggerTask(),
+      TriggerTask(),
+      TriggerTask(),
+    )
+
+    triggeredTasks.foreach(taskQueue.add)
+    taskQueue.start()
+
+    eventually {
+      triggeredTasks(0).triggered shouldBe true
+      triggeredTasks(1).triggered shouldBe true
+      triggeredTasks(2).triggered shouldBe true
+    }
+
+    taskQueue.stop()
+  }
+
+  it should "process other tasks if one task fails" in {
+    val taskQueue = new ActiveTaskQueue[Any]()
+    val triggeredTasks = List(
+      TriggerTask(),
+      TriggerTask(shouldFail = true),
+      TriggerTask(),
+    )
+
+    triggeredTasks.foreach(taskQueue.add)
+    taskQueue.start()
+
+    eventually {
+      triggeredTasks(0).triggered shouldBe true
+      triggeredTasks(2).triggered shouldBe true
+    }
+
+    taskQueue.stop()
+  }
+
+  it should "process incoming tasks" in {
+    val taskQueue = new ActiveTaskQueue[Any]()
+    val triggeredTasks = List(
+      TriggerTask(),
+      TriggerTask(),
+      TriggerTask(),
+    )
+
+    // starting the taskQueue before adding tasks
+    taskQueue.start()
+
+    triggeredTasks.foreach(taskQueue.add)
+
+    eventually {
+      triggeredTasks(0).triggered shouldBe true
+      triggeredTasks(1).triggered shouldBe true
+      triggeredTasks(2).triggered shouldBe true
+    }
+
+    taskQueue.stop()
+  }
+
+  it should "not fail on an empty task queue" in {
+    val taskQueue = new ActiveTaskQueue[Any]()
+    taskQueue.start()
+    taskQueue.stop()
+  }
+
+  "stop" should "cancel all pending tasks" in {
+    val taskQueue = new ActiveTaskQueue[Any]()
+
+    // this mocked task will stop the ActiveTaskQueue when it gets processed
+    // therefore all tasks following it will get skipped
+    val stopTask = mock[Task[Any]]
+    (stopTask.run _)
+      .expects()
+      .onCall(() => {
+        taskQueue.stop()
+        Success(())
+      })
+
+    val triggeredTasks = List(
+      TriggerTask(),
+      stopTask,
+      TriggerTask(),
+      TriggerTask(),
+    )
+
+    triggeredTasks.foreach(taskQueue.add)
+    taskQueue.start()
+
+    eventually {
```

Comment:
* Because of the use of the `IntegrationPatience` mixin, `eventually` will retry the statements every 150 milliseconds, with a timeout of 15 seconds. ([docs](http://doc.scalatest.org/1.8/org/scalatest/concurrent/Eventually.html))
* That is a good point, in the latest commit I made sure there is a one second delay between the `stopTask` triggering and the tasks 2 and 3 being checked. Of course this can still lead to false positives in very rare occasions, but I don't know how to make this test perfect.

Reply:
I guess this is good enough. In fact the wait is not even necessary. The fact that you  call stop on the processing thread of the queue guarantees that the task list is cleared and the stop item added before it the while-loop gets a chance to take another item. This would of course change if we ever changed the execution context to a thread pool or something.